contracts/starknet/Authenticators/EthSig.cairo

```diff
@@ -0,0 +1,370 @@
+# E.g.:
+# - `0x123` will return `3`
+# - `0x1` will return `1`
+# - `0xa3b1d4` will return `6`
+# Notice: Will not work for `0x0` (will return `0` for `0x0` instead of `1`).
+func get_base16_len{range_check_ptr}(num : Uint256) -> (res : felt):
+    let (is_eq) = uint256_eq(num, Uint256(0, 0))
+    if is_eq == 1:
+        return (0)
+    else:
+        # Divide by 16
+        let (divided, _) = uint256_unsigned_div_rem(num, Uint256(16, 0))
+
+        let (res_len) = get_base16_len(divided)
+        return (res_len + 1)
+    end
+end
+
+# Computes `base ** exp` where `base` and `exp` are both `felts` and returns the result as a `Uint256`.
+func u256_pow{range_check_ptr}(base : felt, exp : felt) -> (res : Uint256):
+    alloc_locals
+
+    if exp == 0:
+        # Any number to the power of 0 is 1
+        return (Uint256(1, 0))
+    else:
+        # Compute `base ** exp - 1`
+        let (recursion) = u256_pow(base, exp - 1)
+
+        let (uint256_base) = felt_to_uint256(base)
+
+        # Multiply the result by `base`
+        let (res, overflow) = uint256_mul(recursion, uint256_base)
+
+        with_attr error_message("Overflow happened"):
+            let (no_overflow) = uint256_eq(overflow, Uint256(0, 0))
+            assert no_overflow = 1
+        end
+
+        return (res)
+    end
+end
+
+# Right pads `num` with `0` to make it 32 bytes long.
+# E.g:
+# - right_pad(0x1)  -> (0x0100000000000000000000000000000000000000000000000000000000000000)
+# - right_pad(0xaa) -> (0xaa00000000000000000000000000000000000000000000000000000000000000)
+func pad_right{range_check_ptr}(num : Uint256) -> (res : Uint256):
+    let (len_base16) = get_base16_len(num)
+
+    let (_, rem) = unsigned_div_rem(len_base16, 2)
+    if rem == 1:
+        # Odd-length: add one (a byte is two characters long)
+        tempvar len_base16 = len_base16 + 1
+    else:
+        tempvar len_base16 = len_base16
+    end
+
+    let base = 16
+    let exp = 64 - len_base16
+    let (power_16) = u256_pow(base, exp)
+
+    # Left shift
+    let (low, high) = uint256_mul(num, power_16)
+
+    with_attr error_message("overflow?"):
+        assert high.low = 0
+        assert high.high = 0
+    end
+
+    return (low)
+end
+
+func authenticate_proposal{
+    syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr, bitwise_ptr : BitwiseBuiltin*
+}(
+    r : Uint256,
+    s : Uint256,
+    v : felt,
+    salt : Uint256,
+    target : felt,
+    calldata_len : felt,
+    calldata : felt*,
+):
+    alloc_locals
+
+    # Proposer address should be located in calldata[0]
+    let proposer_address = calldata[0]
+
+    # Ensure proposer has not already used this salt in a previous action
+    let (already_used) = salts.read(proposer_address, salt)
+
+    with_attr error_message("Salt already used"):
+        assert already_used = 0
+    end
+
+    # Execution_hash should be in calldata[1] and calldata[2]
+    let execution_hash = Uint256(calldata[1], calldata[2])
```

Comment:
We now dont store a separate execution hash, this PR: https://github.com/snapshot-labs/sx-core/pull/194